engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/PartSubGraphCompiler.scala

```diff
@@ -251,6 +251,21 @@ class PartSubGraphCompiler(protected val classLoader: ClassLoader,
         val returnType = computeReturnType(objWithMethod, params)
         val typingResult = ServiceTypingResult(returnType, params.map(p => p.name -> p.typingInfo).toMap)
         (compiledgraph.service.ServiceRef(n.id, invoker, params), typingResult)
+      }.andThen{ case (serviceRef, serviceTypingResult) =>
+        val errors = objWithMethod.validationMethod.map{ validationMethod =>
+          val params = serviceRef.parameters
+            .map{ case evaluatedparam.Parameter(name, expression, returnType, _) => name -> (expression, returnType) }
+            .groupBy(_._1)
+            .mapValues(values => values.head._2)
+          validationMethod(params)
+        }.getOrElse{
+          List.empty
+        }
+        if (errors.isEmpty) {
+          valid(serviceRef -> serviceTypingResult).toValidatedNel
+        } else {
+          invalid(CustomValidationErrors(errors, nodeId.id)).toValidatedNel
+        }
```

Comment:
WDYT about leaving `ObjectWithMethodDef` and `DefinitionExtractor` as it was before, in favour of doing validation directly here in this class in some private method:

```
  private def customValidation(objWithMethod: ObjectWithMethodDef, parameters: List[compiledgraph.evaluatedparam.Parameter]): ValidatedNel[CustomValidationErrors, Unit] = ???
``` 
(it would be similar to `private def computeReturnType`)

and then applying this validation after `compileEagerObjectParameters` like that:

```
    val validatedServiceWithTypingResult = service.andThen { objWithMethod =>
      expressionCompiler.compileEagerObjectParameters(objWithMethod.parameters, n.parameters, ctx).map { params =>
        val invoker = createServiceInvoker(objWithMethod)
        val returnType = computeReturnType(objWithMethod, params)
        val typingResult = ServiceTypingResult(returnType, params.map(p => p.name -> p.typingInfo).toMap)
        (compiledgraph.service.ServiceRef(n.id, invoker, params), typingResult)
      }.andThen {
        case compiled@(compiledgraph.service.ServiceRef(n.id, _, params), _) => customValidation(objWithMethod, params).map(_ => compiled)
      }
    }
```

?

Reply:
That would make sens, but in the end I've done Your initial idea (reusing ServiceReturningType).